### PR TITLE
Community join requests

### PR DESCRIPTION
**Pull request by wushenshen**
addresses #499

### CODE REVIEW

**Comment by wushenshen** [comment]
@tibetsprague This addresses items 1, 2, and 4 from the issue - I chose to set all the existing rows to rejected (status = 2) rather than delete them, feels safer. Thoughts on this migration?

File: migrations/20200707214505_add_join_requests_status.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    table.index(['community_id', 'status'])
+  })
+
+  const now = new Date().toISOString()
+  await knex.raw(`UPDATE "join_requests" set status = 2, updated_at = '${now}'`)
+};
+  
+exports.down = async function(knex, Promise) {
+  await knex.schema.table('join_requests', table => {
+    table.dropIndex(['community_id', 'status'])
```
Comment (tibetsprague) [inline migrations/20200707214505_add_join_requests_status.js:14]: dropping the column will automatically drop the index I believe but this is also fine

File: api/graphql/mutations/join_request.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+export async function joinCommunity (communityId, userId) {
+  const user = await User.find(userId)
+  const community = await Community.find(communityId)
+
+  return GroupMembership.forPair(user, community, {includeInactive: true}).fetch()
+    .then(existingMembership => {
+      if (existingMembership) return existingMembership.get('active')
+        ? existingMembership
+        : existingMembership.save({active: true}, {patch: true}).then(membership => membership)
+      if (!!community) return user.joinCommunity(community).then(membership => membership)
```
Comment (tibetsprague) [inline api/graphql/mutations/join_request.js:10]: kind of seems like this check for the community should happen sooner in this function? not a big deal but why do anything if you cant find the user or the community?

File: api/graphql/mutations/join_request.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+export async function createJoinRequest (communityId, userId) {
+  return JoinRequest.create({
+    userId: userId,
+    communityId,
+  })
+  .then(request => ({ request }))
+}
+
+export async function acceptJoinRequest (joinRequestId, communityId, userId, moderatorId) {
```
Comment (tibetsprague) [inline api/graphql/mutations/join_request.js:23]: do we need to check somewhere that this is being called by the userId only?

File: api/graphql/mutations/join_request.js (hunk 2 above)
Comment (wushenshen) [reply api/graphql/mutations/join_request.js]: @tibetsprague good point - I pulled this line out and removed the rest - `user.JoinCommunity` leads to `group.addMembers` which covers the case of an existing membership

File: api/graphql/mutations/join_request.js (hunk 3 above)
Comment (wushenshen) [reply api/graphql/mutations/join_request.js:17]: I added a couple checks for both user and community in the joinCommunity mutation, is that what you had in mind?

**Pull request merged**